### PR TITLE
Spanner gapic migration

### PR DESCRIPTION
**Pull request by yihanzhen**
- Spanner: LRO (`createInstance`, `updateInstance`, `createDatabase`, `updateDatabaseDdl`) methods are changed to return `OperationFuture` instead of `Operation`.
- Spanner: `SpannerOptions.Builder` now takes a `TransportChannelProvider` instead of a `RpcChannelFactory`.
- Spanner: `SpannerOptions.Builder` now takes a `GrpcInterceptorProvider` which provides interceptors to intercept grpc calls.

### CODE REVIEW

**Comment by hzyi-google** [comment]
PTAL @snehashah16 @nithinsujir . Most of the changes were reviewed before. The only new changes I made were within the [examples](https://github.com/googleapis/google-cloud-java/pull/3881/commits/6f6070a6306c9fc0d8a95c59396995c6170d32ec#diff-352be97528eeec92091c254b3a76966dL26) because of the surface change.

**Comment by hzyi-google** [comment]
The tests for windows did pass but it seems the build failed due to: 
`/dev/fd/63: line 1: syntax error near unexpected token 'newline'`
`/dev/fd/63: line 1: '<html>'`
@chingor13 @JesseLovelace Do you know what may have gone wrong?

**Comment by chingor13** [comment]
It looks like maybe the codecov script failed. It does fetch a bash script from codecov.io and pipes it to bash so if that curl failed, then you'd see a bash syntax error.

File: google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GrpcSpannerRpc.java [hunk 1]
```diff
@@ -104,581 +104,581 @@
...
+  //       }
+
+  //       @Override
+  //       public void cancel(@Nullable String message, @Nullable Throwable cause) {
+  //         callLogger.logfmt("Cancelled with message %s", message);
+  //         super.cancel(message, cause);
+  //       }
+  //     };
+  //   }
+  // }
```
Comment (snehashah16) [inline google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GrpcSpannerRpc.java]: y is this file commented out ?

File: google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/testing/RemoteSpannerHelper.java [hunk 2]
```diff
@@ -95,13 +97,16 @@ public String getUniqueDatabaseId() {
...
+      OperationFuture<Database, CreateDatabaseMetadata> op =
         client.getDatabaseAdminClient().createDatabase(instanceId.getInstance(), dbId, statements);
-    op = op.waitFor();
-    Database db = op.getResult();
-    logger.log(Level.FINE, "Created test database {0}", db.getId());
-    dbs.add(db);
+      Database db = op.get();
+      logger.log(Level.FINE, "Created test database {0}", db.getId());
+      dbs.add(db);
     return db;
```
Comment (snehashah16) [inline google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/testing/RemoteSpannerHelper.java]: fix indent

**Comment by snehashah16** [comment]
could u please fix the codacy-bot comments too

also, if it does not cause too much noise, could u format the files too. there seems to be inconsistent formatting throughout the files.

File: google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GrpcSpannerRpc.java (hunk 1 above)
Comment (hzyi-google) [reply google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GrpcSpannerRpc.java]: deleted.

File: google-cloud-clients/google-cloud-spanner/src/test/java/com/google/cloud/spanner/DatabaseAdminClientImplTest.java [hunk 3]
```diff
@@ -82,45 +85,52 @@ public void getDatabase() {
...
-    Operation<Void, UpdateDatabaseDdlMetadata> op =
-        client.updateDatabaseDdl(INSTANCE_ID, DB_ID, ddl, opId);
-    assertThat(op.getName()).isEqualTo(opName);
+    OperationFuture<Empty, UpdateDatabaseDdlMetadata> originalOp =
+        OperationFutureUtil.immediateOperationFuture(
+            originalOpName, Empty.getDefaultInstance(), UpdateDatabaseDdlMetadata.getDefaultInstance());
+    
+    String newOpName = DB_NAME + "/operations/newop";
+    String newOpId = "newop";
+    OperationFuture<Empty, UpdateDatabaseDdlMetadata> newop =
```
[the remark this answers is not in the document]
Comment (hzyi-google) [reply google-cloud-clients/google-cloud-spanner/src/test/java/com/google/cloud/spanner/DatabaseAdminClientImplTest.java]: removing the `newop` would make this test a bit confusing. @snehashah16 Are you fine with keeping it here?

**Comment by pongad** [comment]
Generally LGTM. I'm not sure we should bother with Codacy however; I find a lot of the suggestions questionable. @hzyi-google FYI, you can use [this script](https://github.com/google/google-java-format/blob/master/scripts/google-java-format-diff.py) to format only changed lines.

**Comment by hzyi-google** [comment]
@pongad that is what I used :)

**Pull request merged**